### PR TITLE
#161966928 Implement Article Search and Filter Feature

### PR DESCRIPTION
**Pull request by ThaDeveloper**
#### What does this PR do?
- Implement article search and filter functionalities.
#### Description of Task to be completed?
- Implement search by keywords.
- Implement filter by author, title and tags.
- Implement tests
#### How should this be manually tested?
1. `git clone -b ft-search-and-filter-articles-161966928 https://github.com/andela/ah-infinity-stones` and `$ cd ah-infinity-stones`
2. Create and activate virtual env
- `virtualenv  -p python3 venv`
- `$ source venv/bin/activate`
3. `$ pip3 install -r requirements.txt` to install the dependencies
4. Setup environment variables. Copy all from below and paste them into the .env file. Update the values to your choice.
```export DATABASE_NAME='ah_dev'
export DATABASE_USER='user'
export DATABASE_PASSWORD=pass'
export HOST='localhost'
export PORT='5432'
export DATABASE_TEST='ah_test'
export EMAIL_HOST_USER='username'
export EMAIL_HOST_PASSWORD='pass'
export SECRET_KEY='key-is-secret'
```
Activate the env variables by `$source .env`.
5. Run migrations: `$ python3 manage.py migrate`
6. Run app - `$ python3 manage.py runserver`
7. Acess `http://127.0.0.1:8000/api/users/` through [Postman](https://www.getpostman.com/), provide inputs in the following format and create account.
```
{ "user": {
"username": "username44",
"email": "username@user.com",
"password": "@Password321"	
}
}
```
Check your email and activate account. Use the registration token to acess the following endpoints:
- `http://127.0.0.1:8000/api/articles` - Post an article. Try posting several articles to filter
-  Search and filter articles as below:
     `http://127.0.0.1:8000/api/articles/search/?q=<keyword1, keyword2>` - Search by keywords, one or more.
      `http://127.0.0.1:8000/api/articles/search/?tag=<author>` - Filter by `tag,author,title`

8. Run tests`python3 manage.py test authors.apps.articles.tests.test_views.ArticleSearchTest`
#### What are the relevant pivotal tracker stories?
[#161966928](https://www.pivotaltracker.com/story/show/161966928)
#### Screenshots
screenshot 2018-12-19 at 15 55 27
screenshot 2018-12-19 at 15 56 03

### CODE REVIEW

**Review by PromasterGuru — CHANGES_REQUESTED** [review]
Hey @ThaDeveloper , Kindly rebase develop and resolve the merge conflicts.

**Comment by ThaDeveloper** [comment]
@PromasterGuru Will fix this, they have just arisen after https://github.com/andela/ah-infinity-stones/pull/40 has been merged to develop.

**Review by codeplus254 — COMMENTED** [review]
Great work @ThaDeveloper . I was just wondering whether it is possible to filter search results? Supposing a user got a 100 hits on a search query can they filter the results? 👍

**Comment by ThaDeveloper** [comment]
@codeplus254 Yes, it's possible. So let's say you have filtered by `author` and you get a 100 hits, you can add another filter params like `title` or `tag` to narrow it down to a few hits.

File: authors/apps/articles/models.py [hunk 1]
```diff
@@ -1,7 +1,7 @@
 import uuid
-
 from django.db import models
 from taggit.managers import TaggableManager
+from django.utils.text import slugify
```
Comment (r-wambui) [inline authors/apps/articles/models.py]: I thought this was already on develop why include another one?

Comment (r-wambui) [reply authors/apps/articles/models.py]: Check line 6 of the same file

Comment (ThaDeveloper) [reply authors/apps/articles/models.py]: I may have accepted changes wrongly when rebasing. Let me fix it right away.

**Pull request merged**